---
title: ''
output: 
  html_document: default
  pdf_document: default
---
  
<br></br>  
<center> <h3>
<p><b><span style="color:#337ab7">**Survival Genie**</span></b></p></h3></center>
<center> <h4>
<p><b><span style="color:#337ab7">A web portal for single-cell data, gene-ratio, and cell composition-based survival analyses</span></b></p></h4></center>
<br></br> 
<h5><span style="color:#337ab7">**Motivation**:</span> The genomics data-driven identification of gene signatures, and pathways has been routinely explored for predicting cancer survival and making decisions related to targeted treatments. A large number of packages and tools have been developed to correlate expression or mutations of a gene to clinical outcome, but lack on performing such analysis  based on pathways, gene sets, and gene ratios. Furthermore, in this single-cell omics era, the cluster markers from cancer single-cell transcriptomics studies remains an underutilized prognostic option. Additionally, no bioinformatics online tool evaluates associations between the enrichment of canonical cell types and survival outcome across cancers.</h5>

<i><h5>Here we developed Survival Genie, an interactive R shiny tool to  correlate gene sets, pathways, cellular enrichment and single cell signatures to clinical outcome to assist in developing next generation prognostic and therapeutic biomarkers.</h5></i>

<br></br>
<center>![Survival Genie Overview](./www/overview.png)</center>
<br></br>

<br></br>
<h5>
<span style="color:#337ab7">**Significance**:</span> The analytical options and comprehensive collection of cancer datasets makes Survival Genie a unique resource to correlate gene sets, pathways, cellular enrichment and single cell signatures to  clinical outcome to assist in developing next generation prognostic and therapeutic biomarkers.
</h5>
<br></br>

<h5>
<span style="color:#337ab7">**How it works:**</span>
User selects the analysis type (e.g., single cell data or gene sets), select the inputs, and select the dataset (e.g., TCGA-LAML, TARGET-AML), and partitioning method (i.e., mean, median, quartile, cutp) to determine effect of their levels on patient survival outcomes.
</h5>

<br></br>

<h5>
<span style="color:#337ab7">**Input:**</span> Users are able to select a molecular input from a variety of options:
</h5>

* <h5><i>Gene-based<h5></i> 

  * <h5>single-cell RNA-seq data containing all markers from each cluster</h5>

  * <h5>list of genes representing a set or signature</h5>

  * <h5>input of two-genes</h5>

  * <h5>single-gene input</h5>
  
* <h5><i>Cell-based<h5></i> 

  * <h5>proportion of tumor-infiltrating lymphocytes (TILs) using CIBERSOFT method</h5>
  
  * <h5>percentage of tumor-infiltrating lymphocytes (TILs) from H&E digital images</h5>

* <h5><i>Profile-based<h5></i> 
  * <h5>gene expression profiles  (e.g., derived from weighted gene co-expression network analysis)</h5>

* <h5><i>Mutation-based<h5></i> 
  * <h5>tumor mutation burden (TMB) estimated by the total number of non-synonymous, exonic or all somatic mutations</h5>

<br></br>
<h5>
<span style="color:#337ab7">**Output:**</span> The tool provide comprehensive results including,
</h5>

* <h5>Box plot distribution of high and low sample groups</h5>

* <h5>Correlation of molecular profile and cell composition</h5>

* <h5>Kaplan-Meier plot with log-rank p-value</h5>

* <h5>Hazard-Ratio plot with wald-test p-value</h5>

* <h5>Univariate Cox-Regression survival analysis table</h5>


<br></br>
<h5>
<span style="color:#337ab7">**Datasets**:</span> Survival Genie contains,
</h5>

* <h5>53 [datasets](Datasets.html) of 27 distinct malignancies from 11 different cancer [programs](Programs.html) for both adult and pediatric cancers</h5>

<br></br>
<h5>
<span style="color:#337ab7">**Data collection and processing**:</span>
</h5>

* <h5>The clinical and omics data was downloaded from the [GDC data portal](https://portal.gdc.cancer.gov/) using The GenomicDataCommons Bioconductor R package. Only patients of available clinical survival and tumor mRNA-seq normalized FPKM expression or Whole Exome Seq derived somatic variants were processed.</h5>

* <h5>The H&E digital image-based quantification of Tumor-infiltrating lymphocytes (TILs) was downloaded from the [Cancer Imaging Archive](https://www.cancerimagingarchive.net/) for available 13 TCGA datasets. The proportion of six (LM6 signature) and 22 (LM22 signature) infiltrating immune cell types in tumors were estimated using [CIBERSORT](https://cibersort.stanford.edu/) algorithm.</h5>

* <h5>Gene expression profile (GEP) scores are calculated as weighted sum of the normalized FPKM expression values of the gene signature for each tumor. In the example data, the weightings of each gene in the signature are obtained from (claim # 21c) as published in the Patent filed under [“WO201609437”](https://patentscope.wipo.int/search/en/detail.jsf?docId=WO2016094377&tab=PCTCLAIMS). Otherwise, the GEP scores are computed based on the user-uploaded gene signature weightings.</h5>

* <h5>The somatic mutations reported by Mutect2 pipeline from the [GDC data portal](https://portal.gdc.cancer.gov/) were downloaded and used to estimate the Tumor Mutation Burden (TMB) by the total number of non-synonymous, exonic, and all somatic mutations per Mb.</h5>

<br></br>
<h5>
<span style="color:#337ab7">**Survival analysis**:</span>
</h5>

* <h5>Overall survival (OS) and Event-free survival (EFS) statistical analysis is performed using the ‘survival’ R package.</h5>

* <h5>Kaplan Meier survival curves are used to estimate the OS/EFS using survfit function and a log-rank test is done to compute differences in OS/EFS between the defined high and low groups. Univariate analysis with Cox proportional hazards regression model is performed on the data set using coxph function. </h5>

* <h5>The results are considered significant if the p-values from log rank and Wald test were below 0.05.</h5>


